### DISCUSSION TITLE
Anyone added support for reading NRT metadata to Cecil?

### DISCUSSION THREAD

**Discussion by jpobst** [jbevain/cecil#809] [Q&A]
We are facing a need to be able to determine the nullability status of type references in Cecil.  Unfortunately this seems governed by a [non-trivial encoding format](https://github.com/dotnet/roslyn/blob/main/docs/features/nullable-metadata.md).

Has anyone attempted to implement this using Cecil?

It would be nice to get one canonical implementation either in the base Cecil library or as Rocks extension methods.

**Comment by jbevain** [comment]
Hey @jpobst, i'm afraid I'm not aware of an existing implementation.

**Comment by Zastai** [comment]
It's not super trivial, indeed, especially with the multiple levels of specification, as well as the fact that one attribute will have a list of flags to be resolved to various types involved in a context (type parameters, return types, argument types, etc.).

If you ignore `[NullableContext]`, it's reasonably doable (as long as you make sure to know which flag applies to which type), but full support is tricky indeed. I have some helper methods [here](https://github.com/Zastai/Zastai.Build.ApiReference/blob/69f785f0435c127562d3cd5e3e1cac8f4ba1f386/Zastai.Build.ApiReference/CecilUtils.cs#L168) and [here](https://github.com/Zastai/Zastai.Build.ApiReference/blob/69f785f0435c127562d3cd5e3e1cac8f4ba1f386/Zastai.Build.ApiReference/CecilUtils.cs#L48-L70) but I gave up trying to get it right in the bulk of cases (I only use it for a generic type constraint).

**Reply by jpobst** [reply]
Here is what I ended up with:
https://github.com/xamarin/java.interop/blob/main/src/Java.Interop.Tools.Cecil/Java.Interop.Tools.Cecil/NullableReferenceTypesRocks.cs

**Reply by Zastai** [reply]
If I read that correctly, you do handle the `NullableContext` hierarchy, but only ever get the first nullable flag. So you don't handle, say, a method return type of `Dictionary<string, object?>?` - you would only get the info about the `Dictionary`, not the `string` or the object.

I'll have another crack at it; main thing I may need to do is to pass the context (method or type) to anything that might need to determine nullability.
